File: mutable_self/src/unparser.rs
//! An example of a static analysis.
use std::fmt::Write;

use crate::ast::{Expr, Factor, FactorBinaryOp, Stmt, TermBinaryOp};

#[derive(Debug, Default)]
pub(crate) struct Unparser {
    indent: usize,
    out: String,
}

const INDENT_WIDTH: usize = 2;

impl Unparser {
    pub(crate) fn output(&self) -> &str {
        &self.out
    }

    pub(crate) fn unparse_stmts(&mut self, stmts: &[Stmt]) -> Result<(), std::fmt::Error> {
        for stmt in stmts {
            self.unparse_stmt(stmt)?;
        }
        Ok(())
    }

    fn unparse_stmt(&mut self, stmt: &Stmt) -> Result<(), std::fmt::Error> {
        match stmt {
            Stmt::Let(name, expr) => {
                write!(
                    self.out,
                    "{:indent$}let {name} = ",
                    "",
                    indent = self.indent * INDENT_WIDTH,
                    name = name
                )?;
                self.unparse_expr(expr)?;
                writeln!(self.out)?;
            }
            Stmt::Fun(name, fun) => {
                write!(
                    self.out,
                    "{:indent$}fun {name}(",
                    "",
                    indent = self.indent * INDENT_WIDTH,
                    name = name
                )?;
                for (i, param) in fun.params.iter().enumerate() {
                    if i > 0 {
                        write!(self.out, ", ")?;
                    }
                    write!(self.out, "{}", param)?;
                }
                write!(self.out, ") = ")?;
                self.unparse_expr(&fun.body)?;
                writeln!(self.out)?;
            }
        }
        Ok(())
    }

    fn unparse_expr(&mut self, expr: &Expr) -> Result<(), std::fmt::Error> {
        match expr {
            Expr::Factor(factor) => self.unparse_factor(factor)?,
            Expr::BinaryOp { op, lhs, rhs } => {
                self.unparse_expr(lhs)?;
                write!(self.out, " ")?;
                self.unparse_term_binary_op(op)?;
                write!(self.out, " ")?;
                self.unparse_expr(rhs)?;
            }
        }
        Ok(())
    }

    fn unparse_term_binary_op(&mut self, op: &TermBinaryOp) -> Result<(), std::fmt::Error> {
        match op {
            TermBinaryOp::Add => write!(self.out, "+"),
            TermBinaryOp::Sub => write!(self.out, "-"),
        }
    }

    fn unparse_factor(&mut self, factor: &Factor) -> Result<(), std::fmt::Error> {
        match factor {
            Factor::Literal(value) => write!(self.out, "{}", value)?,
            Factor::Variable(name) => write!(self.out, "{}", name)?,
            Factor::Group(expr) => {
                write!(self.out, "(")?;
                self.unparse_expr(expr)?;
                write!(self.out, ")")?;
            }
            Factor::BinaryOp { op, lhs, rhs } => {
                self.unparse_factor(lhs)?;
                write!(self.out, " ")?;
                self.unparse_factor_binary_op(op)?;
                write!(self.out, " ")?;
                self.unparse_factor(rhs)?;
            }
            Factor::Call(fun, args) => {
                self.unparse_factor(fun)?;
                write!(self.out, "(")?;
                for (i, arg) in args.iter().enumerate() {
                    if i > 0 {
                        write!(self.out, ", ")?;
                    }
                    self.unparse_expr(arg)?;
                }
                write!(self.out, ")")?;
            }
        }
        Ok(())
    }

    fn unparse_factor_binary_op(&mut self, op: &FactorBinaryOp) -> Result<(), std::fmt::Error> {
        match op {
            FactorBinaryOp::Mul => write!(self.out, "*"),
            FactorBinaryOp::Div => write!(self.out, "/"),
        }
    }
}

File: ocaml/calc/lib/unparser.ml
type ctx = {buf: Buffer.t; indent: int}

let rec unparse_ty (ctx : ctx) (ty : Ast.typ) : unit =
  match ty with
  | Ast.Number ->
      Buffer.add_string ctx.buf "number"
  | Ast.Arrow (params, return_ty) ->
      List.iter
        (fun param ->
          unparse_ty ctx param ;
          Buffer.add_string ctx.buf ", " )
        params ;
      Buffer.add_string ctx.buf " -> " ;
      unparse_ty ctx return_ty

let rec unparse_expr (ctx : ctx) (expr : Ast.expr) : unit =
  match expr with
  | Ast.Factor factor ->
      unparse_factor ctx factor
  | Ast.BinaryOp {op; lhs; rhs} ->
      unparse_expr ctx lhs ;
      let op_str = match op with Ast.Add -> " + " | Ast.Sub -> " - " in
      Buffer.add_string ctx.buf op_str ;
      unparse_expr ctx rhs

and unparse_factor (ctx : ctx) (factor : Ast.factor) : unit =
  match factor with
  | Ast.Literal n ->
      Printf.bprintf ctx.buf "%g" n
  | Ast.Variable id ->
      Buffer.add_string ctx.buf id
  | Ast.Group inner_expr ->
      Buffer.add_string ctx.buf "(" ;
      unparse_expr ctx inner_expr ;
      Buffer.add_string ctx.buf ")"
  | Ast.BinaryOp {op; lhs; rhs} ->
      Buffer.add_string ctx.buf "(" ;
      unparse_factor ctx lhs ;
      let op_str = match op with Ast.Mul -> " * " | Ast.Div -> " / " in
      Buffer.add_string ctx.buf op_str ;
      unparse_factor ctx rhs ;
      Buffer.add_string ctx.buf ")"
  | Ast.Call {func; args} ->
      unparse_factor ctx func ;
      Buffer.add_string ctx.buf "(" ;
      let arg_strs =
        List.map
          (fun arg ->
            let buf = Buffer.create 16 in
            let arg_ctx = {buf; indent= 0} in
            unparse_expr arg_ctx arg ; Buffer.contents buf )
          args
      in
      Buffer.add_string ctx.buf (String.concat ", " arg_strs) ;
      Buffer.add_string ctx.buf ")"

let unparse_param (ctx : ctx) (param : Ast.param) : unit =
  Buffer.add_string ctx.buf param.name ;
  Buffer.add_string ctx.buf ": " ;
  unparse_ty ctx param.ty

let unparse_stmt (ctx : ctx) (stmt : Ast.stmt) : unit =
  match stmt with
  | Ast.Let (id, ty, expr) ->
      Printf.bprintf ctx.buf "%slet %s: " (String.make ctx.indent ' ') id ;
      unparse_ty ctx ty ;
      Buffer.add_string ctx.buf " = " ;
      unparse_expr ctx expr ;
      Buffer.add_string ctx.buf "\n"
  | Ast.Fun (id, func) ->
      Printf.bprintf ctx.buf "%sfun %s(" (String.make ctx.indent ' ') id ;
      let num_params : int = List.length func.params in
      List.iteri
        (fun i param ->
          unparse_param ctx param ;
          if i < num_params - 1 then Buffer.add_string ctx.buf ", " else () )
        func.params ;
      Buffer.add_string ctx.buf "): " ;
      unparse_ty ctx func.return_ty ;
      Buffer.add_string ctx.buf " = " ;
      unparse_expr ctx func.body ;
      Buffer.add_string ctx.buf "\n"

let unparse_stmts (stmts : Ast.stmt list) : string =
  let ctx = {buf= Buffer.create 1024; indent= 0} in
  List.iter (unparse_stmt ctx) stmts ;
  Buffer.contents ctx.buf
